### ISSUE TITLE
Why do I need to have the Lable field in the CreatePredictionEngine<TSrc,> class

### ISSUE THREAD

**Issue by PeterPann23** [dotnet/machinelearning#3744]
Hi,

I am wondering why shoeld TSrc in MLContext.Model.CreatePredictionEngine<TSrc, TDst> contain the Label fields as it should not need it to make a prediction. I get it why the other fields needs to be there but the one fields if's supposed to predict is the one it should be able to do without. 
Omitting the fields in your class however would cause an exception when calling CreatePredictionEngine().

Is this intentional?

**Comment by CESARDELATORRE** [comment]
The InputModel/SampleObservation class (TSrc) needs to have the LABEL column because it is the same class you use for reading from the dataset when training or even when reading from a test dataset to create 'bulk predictions'. Since the PredictionEngine uses the same class as "Prediction Input", it is there.

Of course, when predicting, the LABEL field is not needed and the label can/should be empty.

**Comment by PeterPann23** [comment]
Yes, but I do not need it, I need it if I do mass predictions, sure but if I do 1 prediction I do not. actually when I do a prediction I might just want to provide the float[] if my model doesn't need the fields. 

I feel that the need for the label made it to be mandatory irregardless of a business case.   


Say you have a vector of data that holds the value for say N predictions in N models as at the moment one can make only 1 prediction per model right?

So if I need to make N predictions on the SAME vector I end up making a choice of
a. Several training files where I pick a label field
b. Make a separate data loader for each field
c. Make 1 class with all labels for each predictions

None of these are ideal as I do not need it to start with, in a single prediction, when looking at separation of concerns this looks like a bulk predictions pollutes single predictions

**Comment by glebuk** [comment]
The actual root cause of this behavior is that the training data must have exactly the same schema as during prediction.  The label used by the learner at training downstream, and the column that contains "label" in pipeline do not have to be the same and the label data might participate in several transforms before getting to the learner. As a result there is no automated way to remove that source label form the input and ensure that the overall pipeline to be valid. 
For example, we are training a toy model that parses arithmetic expressions like "a+b=c"  and we want to parse c as label and b and a as features.
 In the past we had made some workaround (notably in NimbusML)  by adding an "optional column transform" that ensures that there is always value passed to all columns, including the one that would become the label.
Basically we have two options: either leave it as is, or add an "optional column transform" to the API.

**Comment by PeterPann23** [comment]
so, are you going to document that the "place holder" is needed. Feel free to close the issue if you feel appropriate.
